vpc/service/internal_apis.go

```diff
@@ -407,3 +409,231 @@ func (vpcService *vpcService) doDetachBranchNetworkInterface(ctx context.Context
+			tracehelpers.SetStatus(err, span)
+			return &vpcapi.ResetSecurityGroupResponse{}, err
+		}
+		enisWithSgToUpdate = append(enisWithSgToUpdate, updatedEni)
+		logger.G(ctx).WithField("resetSg", sgToDelete).Debug("Need to update ", updatedEni, " to default SG ", defaultSecurityGroupID)
+
+	}
+
+	err = resetSgTx.Commit()
+	if err != nil {
+		err = fmt.Errorf("Unable to commit transaction: %w", err)
+		tracehelpers.SetStatus(err, span)
+		return &vpcapi.ResetSecurityGroupResponse{}, err
+	}
+
+	//5. Call AWS to change the interface SG ID to the default SG ID of the VPC
+	logger.G(ctx).WithField("resetSg", sgToDelete).Debug("Call AWS to change the interface SG ID to the default SG ID of the VPC")
+
+	defaultSgIDS := make([]string, 0)
+	defaultSgIDS = append(defaultSgIDS, defaultSecurityGroupID)
+
+	//lock eni no key update and then call the AWS API (still dirty and sg is default), and reset the dirty flag to false
+	for _, branchEni := range enisWithSgToUpdate {
+		err := vpcService.updateENISecurityGroups(ctx, resetSgSession, defaultSgIDS, branchEni)
+
+		if err != nil {
+			err = fmt.Errorf("Unable to modify SG ID on AWS to default for %s: %w", branchEni, err)
+			tracehelpers.SetStatus(err, span)
+			return &vpcapi.ResetSecurityGroupResponse{}, err
+		}
+	}
+	logger.G(ctx).WithField("resetSg", sgToDelete).Debug("Reset SG succeeded.")
+	return &vpcapi.ResetSecurityGroupResponse{}, nil
+}
+
+//Update the SG on AWS with the default SG after verifying that the `dirty_security_groups` is unchanged
+func (vpcService *vpcService) updateENISecurityGroups(ctx context.Context, session *ec2wrapper.EC2Session, defSgS []string, eni string) error {
+	ctx, span := trace.StartSpan(ctx, "updateENISecurityGroups")
+	defer span.End()
+
+	tx, err := vpcService.db.BeginTx(ctx, &sql.TxOptions{})
+	if err != nil {
+		err = errors.Wrap(err, "Cannot begin Tx")
+		tracehelpers.SetStatus(err, span)
+		return err
+	}
+	defer func() {
+		_ = tx.Rollback()
+	}()
+
+	row := tx.QueryRowContext(ctx, `
+SELECT security_groups, dirty_security_groups FROM branch_enis WHERE branch_enis.branch_eni = $1
+FOR NO KEY UPDATE OF branch_enis`, eni)
+
+	var securityGroups []string
+	var dirtySecurityGroups bool
+	err = row.Scan(pq.Array(&securityGroups), &dirtySecurityGroups)
+	if err != nil {
+		err = errors.Wrap(err, "Unable to scan branch_enis for dirty_security_groups")
+		tracehelpers.SetStatus(err, span)
+		return err
+	}
+
+	//Check if anything changed on the SG <-> ENI association
+	if !dirtySecurityGroups && len(securityGroups) == 1 && securityGroups[0] == defSgS[0] {
+		err = tx.Commit()
+		if err != nil {
+			err = errors.Wrap(err, "Cannot commit transaction")
+			tracehelpers.SetStatus(err, span)
+			return err
+		}
+		return nil
+	}
+
+	_, err = session.ModifyNetworkInterfaceAttribute(ctx, ec2.ModifyNetworkInterfaceAttributeInput{
+		NetworkInterfaceId: aws.String(eni),
+		Groups:             aws.StringSlice(defSgS),
+	})
+	if err != nil {
+		_ = tx.Rollback()
```

Comment:
You don't need to explicitly rollback here.

Reply:
Removed.